src/spv/inst-exec.cpp

```diff
@@ -196,6 +200,518 @@ bool Instruction::execute(DataView& data, std::vector<Frame*>& frame_stack, bool
+        makeResult(data, 1, nullptr);
+        Primitive& result = static_cast<Primitive&>(*getValue(1, data));
+
+        // --- Step through traversal
+        result.data.b32 = ray_query.proceed();
+
+        break;
+    }
+    case spv::OpRayQueryGetIntersectionTypeKHR: { // 4479
+        // --- Get the arguments
+        RayQuery& ray_query = static_cast<RayQuery&>(*getFromPointer(2, data));
+        const unsigned intersection = static_cast<Primitive&>(*getValue(3, data)).data.u32;
+
+        // --- Set up the return type
+        makeResult(data, 1, nullptr);
+        Primitive& result = static_cast<Primitive&>(*getValue(1, data));
+
+        result.data.u32 = ray_query.getIntersectionType(intersection);
+
+        break;
+    }
+    case spv::OpReportIntersectionKHR: { // 5334
+        // TODO: update once interpreter supports SBTs
+        // --- Get intersection information
+        std::cout << "WARNING: OpReportIntersectionKHR instruction does not follow specifications at the moment!"
+                  << std::endl;
+        const float hit_t = static_cast<Primitive&>(*getValue(2, data)).data.fp32;
+        const unsigned hit_kind = static_cast<Primitive&>(*getValue(3, data)).data.u32;
+        std::cout << "Intersection shader reported an intersection with hitT = (" << hit_t << ") and hitKind = ("
+                  << hit_kind << ")" << std::endl;
+
+        // --- Set up the return type
+        makeResult(data, 1, nullptr);
+        Primitive& result = static_cast<Primitive&>(*getValue(1, data));
+        result.data.b32 = false;
+
+        // TODO: once intersection shader can be invoked by pipeline, use actual rayTMin and rayTMax.
+        // For now, using constants.
+        const float ray_t_min = 0.0;
+        const float ray_t_max = 10000.0;
+
+        if (hit_t < ray_t_min || hit_t > ray_t_max) {
+            // Intersection is outside of the current ray interval
+            std::cout << "\tRay missed; hitT was not in the range [" << ray_t_min << ", " << ray_t_max << "]"
+                      << std::endl;
+            result.data.b32 = false;
+        } else {
+            // TODO: Invoke any-hit shader.
+            // If ignored by any-hit, return false.
+            // If any-hit rejects it, return false.
+            std::cout << "\t(Not working right now) Invoking any-hit shader..." << std::endl;
+            std::cout << "\tSuccessful any-hit shader" << std::endl;
+            result.data.b32 = true;
+        }
+
+        break;
+    }
+    case spv::OpRayQueryGetRayTMinKHR: { // 6016
+        // --- Get the arguments
+        RayQuery& ray_query = static_cast<RayQuery&>(*getFromPointer(2, data));
+
+        // --- Set up the return type
+        makeResult(data, 1, nullptr);  // location and queue does not matter
+        Primitive& result = static_cast<Primitive&>(*getValue(1, data));
+
+        // --- Store the return value
+        result.data.fp32 = ray_query.getRayTMin();
```

Comment:
This seems wrong. You should not be reaching into the primitive to set its raw value directly. One complication is that the result type may have a different precision than you are expecting, and setting it directly fails to account for that.

Please use Primitive's copyFrom. Granted, I don't have the precision handling code in yet, but that is where it will be.

Please do the same for all other instances. I think it is even safer to do copyFrom for the boolean case just in case the primitive isn't actually a boolean (say in the future Khronos says we can use unsigned ints as the result- copyFrom would handle that).